p2p/sharding.py

```diff
@@ -0,0 +1,215 @@
+    name = "sha"
+    version = 0
+    _commands = [Status, Collations]
+    cmd_length = 2
+
+    logger = logging.getLogger("p2p.sharding.ShardingProtocol")
+
+    def send_handshake(self) -> None:
+        cmd = Status(self.cmd_id_offset)
+        self.logger.debug("Sending status msg")
+        self.send(*cmd.encode([]))
+
+    def send_collations(self, collations: List[Collation]) -> None:
+        cmd = Collations(self.cmd_id_offset)
+        self.logger.debug("Sending {} collations".format(len(collations)))
+        self.send(*cmd.encode(collations))
+
+
+class ShardingPeer(BasePeer):
+    _supported_sub_protocols = [ShardingProtocol]
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.incoming_collation_queue = asyncio.Queue()
+        self.known_collation_hashes = set()
+
+    #
+    # Handshake
+    #
+    async def send_sub_proto_handshake(self) -> None:
+        self.sub_proto.send_handshake()
+
+    async def process_sub_proto_handshake(self,
+                                          cmd: Command,
+                                          msg: protocol._DecodedMsgType) -> None:
+        if not isinstance(cmd, Status):
+            self.disconnect(DisconnectReason.other)
+            raise HandshakeFailure("Expected status msg, got {}, disconnecting".format(cmd))
+
+    #
+    # Receiving Collations
+    #
+    def handle_sub_proto_msg(self, cmd: Command, msg: protocol._DecodedMsgType) -> None:
+        if isinstance(cmd, Collations):
+            self._handle_collations_msg(msg)
+        else:
+            super().handle_sub_proto_msg(cmd, msg)
+
+    def _handle_collations_msg(self, msg: List[Collation]) -> None:
+        self.logger.debug("Received {} collations".format(len(msg)))
+        for collation in msg:
+            try:
+                self.incoming_collation_queue.put_nowait(collation)
+            except asyncio.QueueFull:
+                self.logger.warning("Incoming collation queue full, dropping received collation")
+            else:
+                self.known_collation_hashes.add(collation.hash)
+
+    def send_collations(self, collations: List[Collation]) -> None:
+        self.logger.debug("Sending {} collations".format(len(collations)))
+        for collation in collations:
+            if collation.hash not in self.known_collation_hashes:
+                self.known_collation_hashes.add(collation.hash)
+                self.sub_proto.send_collations(collations)
+
+
+class ShardSyncer(PeerPoolSubscriber):
+    logger = logging.getLogger("p2p.sharding.ShardSyncer")
+
+    def __init__(self, shard: Shard, peer_pool: PeerPool, token: CancelToken = None) -> None:
+        self.shard = shard
+        self.peer_pool = peer_pool
+        self.peer_pool.subscribe(self)
+
+        self.incoming_collation_queue = asyncio.Queue()
+
+        self.collations_received_event = asyncio.Event()
+        self.collations_proposed_event = asyncio.Event()
```

Comment:
Nowhere, I removed them. I've added them initially because peers proposed randomly on their own, but proposing by simply calling a method turned out to be easier to test..